storm-core/src/clj/org/apache/storm/daemon/supervisor.clj

```diff
@@ -425,6 +433,13 @@
          ". State: " state
          ", Heartbeat: " (pr-str heartbeat))
         (shutdown-worker supervisor id)))
+
+    (doseq [storm-id all-downloaded-storm-ids]
```

Comment:
This looks like you removed the check where we used to keep the jars even the assignement goes away. We had another issue where if the topology is rebalanced and the assignment goes away 
and immediately comes back to the same supervisor , it tries re-download the jar and if it happens to be a large jar file than there are chances that supervisor worker gets killed before it even starts as the jar download takes longer time.
IMO , i don't see a reason to remove the storm jars on a rebalance or if the assignement goes away. They should be removed when the topology gets killed.